mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td

```diff
@@ -418,6 +418,7 @@ def LLVM_DIDerivedTypeAttr : LLVM_Attr<"DIDerivedType", "di_derived_type",
     OptionalParameter<"uint64_t">:$sizeInBits,
     OptionalParameter<"uint32_t">:$alignInBits,
     OptionalParameter<"uint64_t">:$offsetInBits,
+    OptionalParameter<"std::optional<unsigned>">:$dwarfAddressSpace,
```

Comment:
Modeling this as an optional optional seems a bit odd. Do you know if there is a sematnic difference between zero and `std::nullopt`? If there is none, it would be nicer to model this as `OptionalParameter<"uint32_t">`.

Reply:
I think the design follows LLVM in this regard. Using an integer would solve the CAPI issue but then it is also a slight difference with regards to LLVM (i.e. when round tripping MLIR would suddenly add extra address space information, which may be ok).

Reply:
In one sense, I believe there is not a semantic difference in terms of what the debuginfo *means* between having no dwarf address space field and having a zero field, but the underlying LLVM implementation will do different things when it gets `std::nullopt` vs 0, IE omitting the field or emitting an attribute specifying `DW_AT_address_space` as 0.  For cases where you aren't targetting a GPU, it makes more sense to omit the attribute even if there is no semantic difference between a lack of attribute and a zero value.  Additionally, I'm not certain that all uses of the `DW_AT_address_space` attribute do ensure that zero is the same as lack of an attribute (IE there may be targets where there is a semantic difference that I'm unaware of).  I mainly modeled this to capture the behaviors of the underlying LLVM implementation, but I'll do a bit of digging to try to see whether it matters.

Reply:
Yeah this is something that's entirely target-defined, so as long as there's a difference in spec between nullopt & 0, it's probably safer to also keep the optional representation in MLIR to avoid any weird bugs down the road?

Reply:
Actually, I thought it didn't, but it appears DWARF ([v5, ch. 7.13](https://dwarfstd.org/doc/DWARF5.pdf)) specifies a common `DW_ADDR_none` as 0. From ch. 2.12:

> The set of permissible values is specific to each target architecture. The value DW_ADDR_none, however, is common to all encodings, and means that no address class has been specified.

I still think that if we want to adopt this, we should probably either remove the optional from the LLVM side, or at least translate 0 as nullopt into LLVM. It does mean we lose one degree of freedom, which is controlling whether the 0 address space actually gets emitted or not in DWARF (it's not clear if that should make a difference).

Reply:
I haven't found anything indicating that lack of DW_AT_address_class is necessarily assumed to be different from a zero value, but I think there is practical value to being able to not emit the attribute when not dealing with GPUs as well as explicitly emit 0 when you are.

Reply:
oh yeah I 100% agree on the part about not emitting the 0 attribute for space saving in the general case. I'm less sure if we need to emit it for certain targets though (it's possible some targets have their quirks).

Though if it was a target-based rule on whether to emit 0, we can encode that in the llvm backend for that target. The frontend should not have to worry about it.

Now are there non-target-based rules? I'm not sure.

Reply:
Oh, @zyx-billy , I hadn't even seen your comments when I made mine, github's UI refreshing is inconsistent between top-level and thread comments...  Yeah, maybe we should remove the optionality entirely.

Reply:
That keeps happening to me, I need to remember to refresh the page each time before I write a comment to check if others have come in...

Reply:
haha yeah me too, it's not always consistent about auto-updating these messages 🤦. OK yeah if you also think it's ok to just use DW_ADDR_none we can do that and simplify away this c-api problem too. It would be nice to change llvm too, though of course that's out of scope for this PR 😂 we can do that later if we want.